### PR TITLE
refactor: Replace AbstractFunction to Delegate

### LINKED ISSUE #113: Remove name as a first parameter to simplify code
This is not exactly an "issue".
But this code could be simplified to only take in the function. AbstractFunction has a Name property which could be used in the code.

This makes it possible to also simplify IExpressionHandler.SetFunction by removing the name parameter. And instead using AbstractFunction.Name

https://github.com/casbin/Casbin.NET/blob/aa4fade2ac39b7315c8796c8937b4720c47bf797/NetCasbin/ManagementEnforcer.cs#L986

### PR DESCRIPTION
**Pull request by Sagilio**
Fixes #113

### CODE REVIEW

File: NetCasbin/Model/FunctionMap.cs [hunk 1]
```diff
@@ -17,15 +17,15 @@ public static FunctionMap LoadFunctionMap()
...
+                FunctionDict = new Dictionary<string, Delegate>()
             };
 
-            map.AddFunction("keyMatch", new KeyMatchFunc());
-            map.AddFunction("keyMatch2", new KeyMatch2Func());
-            map.AddFunction("keyMatch3", new KeyMatch3Func());
-            map.AddFunction("keyMatch4", new KeyMatch4Func());
-            map.AddFunction("regexMatch", new RegexMatchFunc());
-            map.AddFunction("ipMatch", new IPMatchFunc());
+            map.AddFunction("keyMatch",  (Func<string, string, bool>) BuiltInFunctions.KeyMatch);
```
Comment (hsluoyz) [inline NetCasbin/Model/FunctionMap.cs:23]: Can functions like `KeyMatchFunc()` be deleted now?

**Review by hsluoyz — CHANGES_REQUESTED** [review]
Is this a breaking change?

File: NetCasbin/Model/FunctionMap.cs (hunk 1 above)
Comment (Sagilio) [reply NetCasbin/Model/FunctionMap.cs:23]: These can be deleted in the library, but still can be used here now. I think we can retain these functions including AbstractFunction class for the user.
So this is not a breaking change, with no impact on the existing project.

Comment (hsluoyz) [reply NetCasbin/Model/FunctionMap.cs:23]: But you have changed the 2nd arg of AddFunction() from AbstractFunction to Delegate. If the user passes an old AbstractFunction type arg into the new version AddFunction(), I think this doesn't work?

**Review by hsluoyz — APPROVED** [review]

File: NetCasbin/Model/FunctionMap.cs (hunk 1 above)
Comment (Sagilio) [reply NetCasbin/Model/FunctionMap.cs]: AbstractFunction has implemented the implicit operation of converting to Delegate. You can have a look at the latest commit.

**Pull request merged**